### PR TITLE
Minor editorial changes

### PR DESCRIPTION
**Pull request by janaiyengar**

### CODE REVIEW

**Comment by martinthomson** [comment]
Based on this, I would suggest we have a new issue that this can maybe address.

**Comment by janaiyengar** [comment]
PING was always supposed to be reliably delivered... if that changed, I missed it. Do you remember it being different?  (I could be losing it, which is not entirely out of the question.)

**Comment by martinthomson** [comment]
I only have what is written down.  And right now, nothing is.  I think that we used to say that it was reliably sent, but we lost that.  And I'm not sad about it.  Whether or not it was intended to be reliably delivered, I can see a use for not having it so - so why not let the sender decide?

**Comment by janaiyengar** [comment]
I'm actually not opposed to it... I think it is sensible to let the sender decide. Either ways, we should have something in retransmission rules about it.

**Comment by janaiyengar** [comment]
Even then, the statement isn't accurate. An endpoint could decide to use PING as a reliable frame if it chooses, which would make the claim in the current text an overreach.

**Comment by martinthomson** [comment]
Sure, but you could easy change the text to "These frames are do not need to be retransmitted if they are lost."

**Comment by mikkelfj** [comment]
Making the sender decide to retransmit a PING makes it hard to use PING in different parts of the spec precisely and it also makes it hard to use for applications. If a reliable PING is needed it should probably be a separate frame, or it should be reliable and we could a separate PROBE frame.

**Comment by MikeBishop** [comment]
We have a separate PROBE frame; it's called PATH_CHALLENGE.

I think we've said previously that PADDING and PING have identical semantics (to wit, none), except that PING forces the packet to be acknowledged regardless of whether there's anything retransmittable in that packet.  Whether we want it to have that property because it's reliably transmitted or because it's not either ACK or PADDING seems like a philosophical discussion akin to how many angels could dance on the head of a pin.

FWIW, it delivers no information when delivered; I see no point in delivering it reliably.  It elicits an ACK if the packet arrives -- that is a sufficient purpose.

**Comment by mikkelfj** [comment]
I don't care about the number of angels, as long as it isn't implementation dependent.

**Comment by ianswett** [comment]
PING frames are "retransmittable", but there's no need to make them reliable at the transport layer, so I don't think we should specify they are.

For PMTU, if the first packet is lost, I can definitely see the sender not sending another PING.

**Comment by mikkelfj** [comment]
If PING can randomly be retransmitted but does not have to, PING is useless to the application.

If PING is never retransmitted, the application can use that signal for its own purposes such as QoS montoring.

If PING is always retransmitted, the application can use that to make sure a connection stays alive by sending one once in a while.

I don't really care either way, as long is it is one or the other. The the missing alternative can be added later as a new frame if important enough.

**Comment by kazuho** [comment]
@mikkelfj To me that sounds like an API issue rather than a protocol design issue.

Current specification of PING does not require it to be retransmitted. However, the fact does not prohibit the developers of QUIC stack from providing an API that retransmits PING frames to test if the peer is alive.

**Comment by mikkelfj** [comment]
@kazuho It is not an API issue: you cannot design an application layer protocol around PING without knowing about retransmission. For one thing it won't work if you app is too tightly coupled to one implementation, and second, you cannot make assumptions about receiving pings because you don't know if it was a retransmission.

I think some things are fine to leave to the API, such as very special things like partial delivery of streams with gaps, but you couldn't make H3 based on such assumptions, as an example. Having PING unspecified makes it unusable for general application design.

Also, it makes keep alive very error prone since you might incorrectly assume that sending ping will ensure keep alive, but in fact it only work that way for some implementations.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -3035,17 +3035,17 @@ packets (except for PMTU probe packets) SHOULD be sized to fit within the
...
-larger than 1280 bytes. Assuming the minimum IP header size, this results in a
-QUIC maximum packet size of 1232 bytes for IPv6 and 1252 bytes for IPv4. A QUIC
-implementation MAY be more conservative in computing the QUIC maximum packet
-size to allow for unknown tunnel overheads or IP header options/extensions.
+An endpoint SHOULD use Datagram Packetization Layer PMTU Discovery
+({{!DPLPMTUD=I-D.ietf-tsvwg-datagram-plpmtud}}) or implement Path MTU Discovery
+(PMTUD) {{!RFC1191}} {{!RFC8201}} to determine whether the path to a destination
+will support a desired message size without fragmentation.
+
+In the absence of these aforementioned mechanisms, QUIC endpoints SHOULD NOT
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: "these aforementioned" seems redundant.  Either "these mechanisms" (the original) or "the aforementioned mechanisms," though that sounds over-formal.

**Comment by hardie** [comment]
On Thu, Nov 29, 2018 at 12:30 PM MikkelFJ <notifications@github.com> wrote:

> If PING can randomly be retransmitted but does not have to, PING is
> useless to the application.
>
> If PING is never retransmitted, the application can use that signal for
> its own purposes such as QoS montoring.
>
> If PING is always retransmitted, the application can use that to make sure
> a connection stays alive by sending one once in a while.
>

For the heartbeat case, it seems like the application could send a PING
frame and continue to send new PING frames until one is acknowledged (since
they must be acknowledged).  I agree, though, that is a pretty wasteful set
of affairs if you're going to need to send a lot of heartbeats.  I don't
think that strategy for maintain a stable 5 tuple is as necessary with
QUIC, though, so the trade-off looks like this to me:

Case 1:

1)PING MUST be acknowledged but is not re-transmitted, so a lack of
acknowledgement is a clear signal of loss on the path.

2) If you want a heartbeat, you send fresh PINGs until  you get an
acknowledgement of one of them (and count the ones that did not get
acknowledged as lost).

Case 2:

1) PING MUST be acknowledged and MUST be re-transmitted, so a single PING
is sufficient to create a heartbeat for the connection/application.

2) Each re-transmission is a signal of either loss or significant delay
along the path.

If we think heartbeats over lossy connections are going to be common, I'd
go with two.  If we don't, I'd go with one.  I agree that having "might be
retransmitted" is the worst option.

Just my two cents,

Ted

>

**Comment by huitema** [comment]
By the way, the actual text of the PR is fine. It is just the title that is weird.

**Comment by mikkelfj** [comment]
I think we may then need some text to guide application level keep-alive design and explicitly state that PING might not be it.

**Comment by huitema** [comment]
@mikkelfj The way Picoquic does it is through these API:
```
/* Enables keep alive for a connection.
 * If `interval` is `0`, it is set to `idle_timeout / 2`.
 */
void picoquic_enable_keep_alive(picoquic_cnx_t* cnx, uint64_t interval);
/* Disables keep alive for a connection. */
void picoquic_disable_keep_alive(picoquic_cnx_t* cnx);
```
If the API is enabled, the transport will in fact send PING frames, but only if there is not enough application traffic or other traffic to sustain the connection.

**Comment by MikeBishop** [comment]
Sounds like I should probably reword this section in HTTP, then:
> HTTP clients are expected to use QUIC PING frames to keep connections open while there are responses outstanding for requests or server pushes. If the client is not expecting a response from the server, allowing an idle connection to time out is preferred over expending effort maintaining a connection that might not be needed. A gateway MAY use PING to maintain connections in anticipation of need rather than incur the latency cost of connection establishment to servers. Servers SHOULD NOT use PING frames to keep a connection open.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: Yeah, something wasn't sitting right with me after I wrote that out. I'll leave it as it was.

**Pull request merged**